File: 752-open-the-lock/752-open-the-lock.java
class Solution {
    private char scrollForward(char digit){
        if(digit == '9')
            return '0';
        return ++digit;
    }
    private char scrollBackward(char digit){
        if(digit == '0')
            return '9';
        return --digit;
    }
    public int openLock(String[] deadends, String target) {
        var deads = new HashSet<String>(Arrays.asList(deadends));
        if(deads.contains("0000"))
            return -1;
        var begin = new HashSet<String>();
        var end = new HashSet<String>();
        int steps = 0;
            
        begin.add("0000");
        end.add(target);
        
        while(!begin.isEmpty() && !end.isEmpty()){
            if(begin.size() > end.size()){
                var temp = begin;
                begin = end;
                end = temp;
            }
            var nextStates = new HashSet<String>();
            for(String lock : begin){
                if(end.contains(lock))
                    return steps;
                deads.add(lock);
                StringBuilder builder = new StringBuilder(lock);
                for(int i = 0; i < 4; i++){
                    String forward = builder.substring(0, i) + scrollForward(lock.charAt(i)) + builder.substring(i + 1);
                    String backward = builder.substring(0, i) + scrollBackward(lock.charAt(i)) + builder.substring(i + 1);
                    if(!deads.contains(forward))
                        nextStates.add(forward);
                    if(!deads.contains(backward))
                        nextStates.add(backward);
                }
            }
            begin = nextStates;
            steps++;
        }
        
        return -1;
    }
}

File: 752-open-the-lock/752-open-the-lock.cpp
class Solution {
    char moveForward(char digit){
        if(digit == '9')
            return '0';
        return ++digit;
    }
    char moveBackward(char digit){
        if(digit == '0')
            return '9';
        return --digit;
    }
public:
    int openLock(std::vector<std::string>& deadends, std::string target) {
        std::unordered_set<std::string> begin;
        std::unordered_set<std::string> end;
        std::unordered_set<std::string> deads(deadends.begin(), deadends.end());
        if(deads.find("0000") != deads.end())
            return -1;
                                                                          
        begin.insert("0000");
        end.insert(target);
        
        // bidirectional search
        int steps = 0;
        
        while(begin.size() > 0 and end.size() > 0){
            std::unordered_set<std::string> nextStates;
            for(const string& lock : begin){
                if(end.find(lock) != end.end()) // if the two sets intersect at any point
                    return steps;
                deads.insert(lock);
                for(int i = 0; i < 4; i++){
                    std::string forward = lock.substr(0, i) + moveForward(lock[i]) + lock.substr(i + 1);
                    std::string backward = lock.substr(0, i) + moveBackward(lock[i]) + lock.substr(i + 1);
                    if(deads.find(forward) == deads.end())
                        nextStates.insert(forward);
                    if(deads.find(backward) == deads.end())
                        nextStates.insert(backward);
                }
            }
            begin = end;
            end = nextStates;
            steps++;
        }
        
        return -1;
    }
};